test/FullRange.t.sol

```diff
@@ -0,0 +1,413 @@
+        token2.mint(address(this), 2 ** 128);
+
+        manager = new PoolManager(500000);
+
+        vm.record();
+        FullRangeImplementation impl = new FullRangeImplementation(manager, fullRange);
+        vm.etch(address(fullRange), address(impl).code);
+
+        key = PoolKey(Currency.wrap(address(token0)), Currency.wrap(address(token1)), 3000, TICK_SPACING, fullRange);
+        id = key.toId();
+
+        key2 = PoolKey(Currency.wrap(address(token1)), Currency.wrap(address(token2)), 3000, TICK_SPACING, fullRange);
+        id2 = key.toId();
+
+        modifyPositionRouter = new PoolModifyPositionTest(manager);
+        swapRouter = new PoolSwapTest(manager);
+
+        token0.approve(address(fullRange), type(uint256).max);
+        token1.approve(address(fullRange), type(uint256).max);
+        token2.approve(address(fullRange), type(uint256).max);
+        token0.approve(address(modifyPositionRouter), type(uint256).max);
+        token1.approve(address(modifyPositionRouter), type(uint256).max);
+        token2.approve(address(modifyPositionRouter), type(uint256).max);
+        token0.approve(address(swapRouter), type(uint256).max);
+        token1.approve(address(swapRouter), type(uint256).max);
+        token2.approve(address(swapRouter), type(uint256).max);
+        token0.approve(address(manager), type(uint256).max);
+        token1.approve(address(manager), type(uint256).max);
+        token2.approve(address(manager), type(uint256).max);
+    }
+
+    function testBeforeInitializeAllowsPoolCreation() public {
+        vm.expectEmit(true, true, true, true);
+        emit Initialize(id, key.currency0, key.currency1, key.fee, key.tickSpacing, key.hooks);
+        snapStart("initialize");
+        manager.initialize(key, SQRT_RATIO_1_1);
+        snapEnd();
+
+        (, address liquidityToken) = fullRange.poolInfo(id);
+
+        assertFalse(liquidityToken == address(0));
+    }
+
+    function testBeforeInitializeRevertsIfWrongSpacing() public {
+        PoolKey memory wrongKey =
+            PoolKey(Currency.wrap(address(token0)), Currency.wrap(address(token1)), 0, TICK_SPACING + 1, fullRange);
+
+        vm.expectRevert(FullRange.TickSpacingNotDefault.selector);
+        manager.initialize(wrongKey, SQRT_RATIO_1_1);
+    }
+
+    function testInitialAddLiquiditySucceeds() public {
+        manager.initialize(key, SQRT_RATIO_1_1);
+
+        uint256 prevBalance0 = MockERC20(token0).balanceOf(address(this));
+        uint256 prevBalance1 = MockERC20(token1).balanceOf(address(this));
+
+        snapStart("add liquidity");
+        fullRange.addLiquidity(address(token0), address(token1), 3000, 10 ether, 10 ether, address(this), MAX_DEADLINE);
+        snapEnd();
+
+        assertEq(MockERC20(token0).balanceOf(address(this)), prevBalance0 - 10 ether);
+        assertEq(MockERC20(token1).balanceOf(address(this)), prevBalance1 - 10 ether);
+
+        (, address liquidityToken) = fullRange.poolInfo(id);
```

Comment:
personal preference, but I'd gather all variables first (lines 148 + 152) above and then all assertions together underneath

```
```
        (bool owed,) = fullRange.poolInfo(id);
       (, address liquidityToken) = fullRange.poolInfo(id);

        assertEq(MockERC20(token0).balanceOf(address(this)), prevBalance0 - 10 ether);
        assertEq(MockERC20(token1).balanceOf(address(this)), prevBalance1 - 10 ether);
        assertEq(UniswapV4ERC20(liquidityToken).balanceOf(address(this)), 10 ether - 1000);
        assertEq(owed, false);

Reply:
sometimes i need to get a newly updated bool owed after i use the liquidityToken, so i can't put them together all the time, but i will put the owed variable before the asserts

Reply:
just saying that asserts are grouped together after all actions + getters (so nothing should change). sometimes a test might have multiple stages, but `assign beforeParams, action, assign afterParams, assertions` for each stage is a nice way to consistently organize 🤷